### DISCUSSION TITLE
Loading complicated authentication dependencies when using TraceConfiguration and AutoConfigurationCustomizerProvider

### DISCUSSION THREAD

**Discussion by greenhsu123** [open-telemetry/opentelemetry-java-instrumentation#11930] [Q&A]
We are having difficulties making sure various authentication dependencies (own + third party) for TraceConfiguration are being loaded correctly by class loaders when the TraceConfiguration is being passed to AutoConfigurationCustomizerProvider. What is everyone doing in this space?

**Comment by psx95** [comment]
@greenhsu123 could you elaborate on the exact problem are you facing? 
Particularly, what do you mean when you say "TraceConfiguration is being passed to AutoConfigurationCustomizerProvider" ? 

The [AutoConfigurationCustomizerProvider](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizerProvider.java) lets you customize the OpenTelemetry instance used by the Java Agent. I believe your implementation of this interface is what gets loaded by the agent.

**Reply by greenhsu123** [reply]
Hi @psx95 - thank you for getting back to me. 

Below is how we are using the AutoConfigurationCustomizerProvider, and setCredentials() is where we are having issues ensuring various authentication dependencies are being loaded correctly by classloaders. 

```java
    @Override
    public void customize(@NotNull AutoConfigurationCustomizer autoConfigurationCustomizer) {
         
        //com.google.cloud.opentelemetry.trace.TraceConfiguration;
        TraceConfiguration configuration =
            TraceConfiguration.builder()
                .setCredentials()
                .setProjectId()
                .setTraceServiceEndpoint()
                .setDeadline()
                .setFixedAttributes()
                .build();

        SpanExporter traceExporter = TraceExporter.createWithConfiguration(configuration);

        autoConfigurationCustomizer
            .addTracerProviderCustomizer((tpb, ignored) -> tpb
                .addSpanProcessor(
                    BatchSpanProcessor.builder(traceExporter).build())
                .setSampler(Sampler.alwaysOn())
            );
    }
```

**Reply by psx95** [reply]
Hi @greenhsu123,
Thank you for providing the configuration code. In this code, are you seeing any exceptions when you try to run it ?
I noticed that you're not passing in any credentials, which would mean that [TraceExporter would try to use application-default credentials](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/blob/72a456e5e5712ac710818f06e290ceb5ba834e31/exporters/trace/src/main/java/com/google/cloud/opentelemetry/trace/InternalTraceExporter.java#L94) from the `com.google.auth.oauth2`.
IIUC, your issue is that these dependencies are missing at runtime.

If this is the case, one possible issue could be that you're not packaging the dependencies with the extension JAR that you build. 
You could try to build a shadow JAR with all dependencies included using the shadow JAR gradle plugin: 
```groovy
// In your build.gradle
plugins {
	id 'java-library'
	id 'com.github.johnrengelman.shadow'
}

tasks {
	assemble.dependsOn(shadowJar)
}

shadowJar {
	archiveClassifier.set("")
}

jar {
	// Disable standard jar
	enabled = false
}
```

#### Some follow-up Questions:
Are you trying to use [Cloud Trace Exporter for OpenTelemetry](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/tree/main/exporters/trace) with Java Autoinstrumentation agent to export traces to Google Cloud ? 
If so, I think you could use the [AutoConfguration Exporter](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/tree/main/exporters/auto#auto-configuration-for-opentelemetry-in-google-cloud) instead of creating your own extension. It should work right out of the box for you.
Here is an [example](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/tree/main/examples/autoinstrument) that might help you.

In case you're intentionally trying to do this manually, could you explain your reason/use-case behind it ?

**Reply by greenhsu123** [reply]
Hi @psx95, thank you for getting back to me, and apologies that if I weren't clear in the previous code snippet. For the ```TraceConfiguration ``` object, we do pass in  ```com.google.auth.oauth2```, please see below for an updated version. The exception we have been seeing is a ClassCast Exception, where oauth2 object is loaded by 1 classloader but we are trying to use it in another classloader. 
```java
 @Override
    public void customize(@NotNull AutoConfigurationCustomizer autoConfigurationCustomizer) {
         
        //com.google.cloud.opentelemetry.trace.TraceConfiguration;
        TraceConfiguration configuration =
            TraceConfiguration.builder()
                .setCredentials(<OAuth2 object>)
                .setProjectId(<GCP Project ID>)
                .setTraceServiceEndpoint(<GCP Trace Endpoint>)
                .setDeadline(<Deadline>)
                .setFixedAttributes(<attributes>)
                .build(); 
```

In terms of your follow-up question, yes, I am aware of the library, however, we don't want to save credential details to a json file as required by ```GOOGLE_APPLICATION_CREDENTIALS```.

**Reply by psx95** [reply]
Hi @greenhsu123, apologies for the delayed response. 
So from what I understand the issue here is that the `AutoConfigurationCustomizerProvider` can be used to customize the existing OTel SDK, but `TraceExporter` is not in the standard OTel exporters package. So that class is not present at the time this customization runs. 

So what you need to do is first make the Google Cloud's `TraceExporter` be available to the autoconfigured SDKs. This can be done by implementing another SPI - [ConfigurableSpanExporterProvider](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/traces/ConfigurableSpanExporterProvider.java#L12).

Implementing this SPI is exactly what Google Cloud's [AutoConfigurtion Exporter](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/tree/main/exporters/auto#auto-configuration-for-opentelemetry-in-google-cloud) does and so you can directly use the published shaded-jar to make it available.

You can then implement your custom extension to modify the `TraceExporter` made available.

```groovy
// build.gradle
applicationDefaultJvmArgs = [
	// Use the downloaded java agent.
	'-javaagent:agents/otel-agent-2.4.0.jar',
	'-Dotel.service.name=service-name',
        // supply multiple extensions
       -Dotel.javaagent.extensions=libs/gcp_exporter-auto-shaded-0.29.1.jar,extension/custom-auth-extension.jar
       // Configure auto instrumentation.
	'-Dotel.traces.exporter=google_cloud_trace',
	'-Dotel.metrics.exporter=none', // set google_cloud_monitoring if also exporting metrics. 
        // other relevant args...
       ...
]
```
Your code for the auth extension mentioned [above](https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/11930#discussioncomment-10254428) should work then.

I'm not sure about the `ClassCastException` though, but the `ClassNotFoundException` makes sense given the current scenario.
I'd be happy to assist with the `ClassCastException` if you can share the exception and the code which runs into this exception.
